File: SWEA/D3/5607. ［Professional］ 조합/［Professional］ 조합.py
MOD = 1234567891
MAX = 1000000

# 팩토리얼과 역팩토리얼 미리 계산
fact = [1] * (MAX + 1)
inv_fact = [1] * (MAX + 1)

for i in range(1, MAX + 1):
    fact[i] = (fact[i - 1] * i) % MOD

# 역팩토리얼 계산 (페르마의 소정리 이용)
inv_fact[MAX] = pow(fact[MAX], MOD - 2, MOD)
for i in range(MAX, 0, -1):
    inv_fact[i - 1] = (inv_fact[i] * i) % MOD

# 테스트케이스 입력
T = int(input())
for tc in range(1, T + 1):
    N, R = map(int, input().split())
    # 조합: N! / (R! * (N-R)!) mod MOD
    answer = fact[N] * inv_fact[R] % MOD
    answer = answer * inv_fact[N - R] % MOD
    print(f"#{tc} {answer}")

File: SWEA/D3/5607. ［Professional］ 조합/［Professional］ 조합.java
import java.io.*;
import java.util.*;

public class Solution {

    static final long MOD = 1234567891L;
    static long[] fact, invFact;

    static long modPow(long a, long e) {
        long r = 1L;
        while (e > 0) {
            if ((e & 1L) == 1L) r = (r * a) % MOD;
            a = (a * a) % MOD;
            e >>= 1;
        }
        return r;
    }

    static void buildFactorials(int nMax) {
        fact = new long[nMax + 1];
        invFact = new long[nMax + 1];

        fact[0] = 1L;
        for (int i = 1; i <= nMax; i++) {
            fact[i] = (fact[i - 1] * i) % MOD;
        }

        invFact[nMax] = modPow(fact[nMax], MOD - 2); // Fermat inverse
        for (int i = nMax; i >= 1; i--) {
            invFact[i - 1] = (invFact[i] * i) % MOD;
        }
    }

    static long nCr(int n, int r) {
        if (r < 0 || r > n) return 0L;
        return (((fact[n] * invFact[r]) % MOD) * invFact[n - r]) % MOD;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine().trim());
        int[] Ns = new int[T];
        int[] Rs = new int[T];

        int nMax = 0;
        for (int tc = 0; tc < T; tc++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            int r = Integer.parseInt(st.nextToken());
            Ns[tc] = n; Rs[tc] = r;
            if (n > nMax) nMax = n;
        }

        buildFactorials(nMax);

        for (int tc = 0; tc < T; tc++) {
            long ans = nCr(Ns[tc], Rs[tc]);
            sb.append('#').append(tc + 1).append(' ').append(ans).append('\n');
        }

        System.out.print(sb.toString());
    }
}